docs/ttp-dev.md

```diff
@@ -0,0 +1,146 @@
+
+### Metadata
+
+TTP metadata must include the name of the TTP and a description of that TTPs behavior. MITRE ATT&CK IDs are optional but recommended. If the TTP cannot be mapped to MITRE ATT&CK then the `mitre` mapping and it's child mappings should be omitted. 
+
+An example of TTP metadata is shown below.
+```text
+---
+name: Disable system security updates
+description: |
+  This TTP disables the automatic installation of macOS security updates.
+mitre:
+  tactics:
+    - T0005 Defense Evasion
+  techniques:
+    - T1562 Impair Defenses
+  subtechniques:
+    - "T1562.001 Impair Defenses: Disable or Modify Tools"
+
+```
+
+### Arguments
+
+Arguments are defined after the TTP metadata. Arguments are uniquely named, may be strongly typed, and may contain default values. Using an argument in the TTP is accomplished with the argument syntax, `{{ .Args.arg_name }}`. A complete example is shown below.
+
+```text
+args:
+  - name: a_message
+  - name: a_number
+    type: int
+  - name: has_a_default
+    default: this is the default value
+steps:
+  - name: print_args
+    inline: |
+      echo "hi! You passed the message: {{ .Args.a_message }}"
+      echo "You passed the number: {{ .Args.a_number }}"
+      echo "has_a_default has the value: '{{ .Args.has_a_default }}'"
+```
+
+### Steps
+
+Steps are uniquely named blocks of implementation logic which are executed in sequence. Steps help developers organize and manage the complexity of TTPs.
+
+In general, steps will fall into one of the following high-level categories;
+
+- Assessment
+- Shaping
+- Execution
+- Cleanup
+
+Additionally, TTPs may be daisy-chained enabling developers to create complex sequences of TTPs. In doing so, each daisy-chained TTP is represented in the parent TTP as a sub-TTP. We'll see an example of this shortly and some recommendations on developing sub-TTPs as common building blocks.
+
+#### Assessment
+
+It is often necessary for a TTP to test execution requirements, such as whether a necessary environment variable is set, and bail out of the TTP if it is not. In this example, if the `AWS_DEFAULT_REGION` environment variable is not set then the TTP returns exit code 1 and no further blocks are executed.
+
+```text
+steps:
+  - name: ensure-aws-creds-present
+    inline: |
+      set -e
+
+      if [[ -z "${AWS_DEFAULT_REGION}" ]]; then
+          echo "Error: AWS_DEFAULT_REGION must be set."
+          exit 1
+      fi
+    
+	<- snip ->
+```
+
+When creating assessment type blocks, it's preferable to place each test in its own block rather than a single block that tests all prerequisites. This will make your TTPs much easier to maintain as they become more complex.
```

Comment:
```suggestion
When creating assessment type blocks, it's preferable to place each test in its 
own block rather than a single block that tests all prerequisites. This approach
will make your TTPs much easier to maintain as they become more complex.
```